enhancements/installer/agent-late-binding.md

```diff
@@ -0,0 +1,179 @@
+InfraEnv's identity must be encoded in the ISO, and the agent must pass it to
+the service upon registration.  When an agent registers, the backend will
+create the Agent CR in the InfraEnv's namespace, and if the InfraEnv references
+a ClusterDeployment, the Agent will automatically be associated with that
+ClusterDeployment.
+
+The second major change is to the Agent CRD, where the ClusterRef is no longer
+mandatory.  This association may now be deleted or changed dynamically, unless
+the Agent is part of an ongoing cluster installation, in which case the
+installation must first be cancelled.  If the host had started/completed its
+installation, a new Agent Condition would indicate that the user needs to boot
+the host into the Discovery ISO once again.
+
+Agent validations that do not depend on being part of a cluster will run for
+Agents that are both associated with clusters and those that are not.  Examples
+of these include minimum hardware for any role, synced NTP, and registry
+access.  Validations that do depend on being part of a cluster, such as
+connectivity checks, will run only for Agents associated with clusters.
+
+### User Stories
+
+#### Story 1
+
+As an Infrastructure Admin, I want to create a Discovery ISO and use it to boot
+a pool of hosts for use by Cluster Creators to install OpenShift on.
+
+#### Story 2
+
+As an Infrastructure Admin, I want to add Discovery ISOs to hosts as a boot
+option, and then ship the hosts to remote locations where they will be booted
+at a later time for use by Cluster Creators to install OpenShift on.
+
+#### Story 3
+
+As a Cluster Creator, I want to view a collection of available Agents and use
+them for OpenShift cluster creation or expansion.
+
+#### Story 4
+
+As a Cluster Creator, I want to reassign an Agent from one OpenShift cluster to
+another.
+
+### Implementation Details/Notes/Constraints [optional]
+
+- It would be beneficial to move the discovery image management to a separate
+  service at some point, both to allow it to be scaled independently and to
+  reduce the very large scope of the existing service.  The late binding work
+  should be done with this separation in mind, even if the actual separation is
+  an orthogonal task.
+
+- While the Agent and ClusterDeployment resources will be viewed as independent
+  to API users, to avoid a large DB and code migration, internally Agents will
+  still always be associated with clusters in the SQL DB.
+
+- The REST API will be overhauled as well to separate the image and host
+  resources from the cluster resource.
+
+### Risks and Mitigations
+
+## Design Details
+
+### Open Questions [optional]
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+### Graduation Criteria
+
+**Note:** *Section not required until targeted at a release.*
+
+### Upgrade / Downgrade Strategy
+
+Existing CRs should continue to work without modification.
```

Comment:
It's worth noting that downgrade won't be possible. Once we start creating Agents without an associated cluster, we can't go back to a world where Agents require a cluster.